File: evals/mlx-engine-bench/bench.py
#!/usr/bin/env python3
"""MLX engine baseline harness — talks straight to the mounted Rapid-MLX engine.

Usage:
  bench.py --label <label> [--base-url http://127.0.0.1:8090] [--reps 3]
           [--workloads a,b,c] [--nonce <text>] [--temperature 0 | --omit-temperature]

Three fixed workloads, each repeated --reps times:
  a  short chat   ~200-token prompt  -> max 300 tokens out
  b  long prompt  ~32k-token fixed document -> max 200 tokens out (prefill + TTFT)
  c  goose agent  the real goose system prompt + 79 tool schemas (fixtures/, sanitized), a
                  4-turn tool conversation: each turn appends a FIXED assistant tool call and a
                  FIXED tool result, and moves a CHANGING <turn-context> to the request tail the
                  way goose does (moim.rs compose_moim + openai.rs append_turn_context_tail), so
                  the prefix-cache reuse across tool calls is what gets measured.

Every request runs alone (the harness is the only client) and is bracketed by /v1/status and
/metrics snapshots; the deltas are the engine's own counters for that one request.

Why every repetition carries a NONCE as the first line of its prompt: an identical prompt would
be answered from the prefix cache (and, greedy, possibly the response cache), and the cache is
persisted across remounts — so a repeated prompt would measure the cache of a PREVIOUS run, not
the configuration under test. The nonce makes every repetition cold at token ~0; workload (c)
then measures only the reuse INSIDE its own conversation. Pass --nonce explicitly to replay a
conversation on purpose (the unmount -> mount persistence check).

Why temperature 0 and NO seed: greedy decoding makes the output (and so the completion length)
repeatable; a seed would pin Rapid-MLX's MTP depth to max_k and switch off the EV auto-K
controller production runs (scheduler.py:2133-2138 `disable_auto_k=... lane_rng is not None`).
goose itself sends no temperature (the engine's default sampler applies) — --omit-temperature
reproduces that for a production-shaped check.
"""

import argparse
import json
import os
import random
import re
import statistics
import sys
import time
import urllib.request
from datetime import datetime

HERE = os.path.dirname(os.path.abspath(__file__))

# ---------------------------------------------------------------- engine I/O


def http_get(url: str) -> str:
    with urllib.request.urlopen(url, timeout=30) as r:
        return r.read().decode()


def snapshot(base: str) -> dict:
    status = json.loads(http_get(f"{base}/v1/status"))
    metrics = {}
    for line in http_get(f"{base}/metrics").splitlines():
        if not line or line.startswith("#"):
            continue
        m = re.match(r"^([a-zA-Z_:][a-zA-Z0-9_:]*)(\{[^}]*\})?\s+([-+0-9.eEinfNa]+)$", line)
        if not m:
            continue
        name, labels, value = m.group(1), m.group(2) or "", m.group(3)
        if not (name.startswith("rapid_mlx_spec_decode") or name.startswith("rapid_mlx_prefix_cache")):
            continue
        labels = re.sub(r',?model_id="[^"]*"', "", labels)
        try:
            metrics[name + labels] = float(value)
        except ValueError:
            pass
    return {"status": status, "metrics": metrics}


CACHE_KEYS = ["hits", "misses", "tokens_saved", "evictions", "non_trimmable_skips", "load_skipped"]


def deltas(before: dict, after: dict) -> dict:
    cb, ca = before["status"].get("cache", {}), after["status"].get("cache", {})
    cache = {k: ca.get(k, 0) - cb.get(k, 0) for k in CACHE_KEYS if k in ca}
    cache["entry_count_after"] = ca.get("entry_count")
    cache["current_memory_mb_after"] = ca.get("current_memory_mb")
    spec = {}
    for k, v in after["metrics"].items():
        if not k.startswith("rapid_mlx_spec_decode") or "accept_ratio" in k or "k_cost_ms" in k:
            continue
        d = v - before["metrics"].get(k, 0.0)
        if d:
            spec[k.replace("rapid_mlx_spec_decode_", "").replace('family="qwen3.8",', "")] = d
    return {"cache": cache, "spec_decode": spec}


def stream_chat(base: str, body: dict) -> dict:
    body = dict(body, stream=True, stream_options={"include_usage": True})
    req = urllib.request.Request(
        f"{base}/v1/chat/completions",
        data=json.dumps(body).encode(),
        headers={"content-type": "application/json"},
    )
    t0 = time.perf_counter()
    t_first = t_last = None
    usage, finish, req_metrics = None, None, None
    content_chars = reasoning_chars = 0
    tool_calls = []
    with urllib.request.urlopen(req, timeout=None) as r:
        for raw in r:
            line = raw.decode().strip()
            if not line.startswith("data:"):
                continue
            payload = line[5:].strip()
            if payload == "[DONE]":
                break
            chunk = json.loads(payload)
            if chunk.get("usage"):
                usage = chunk["usage"]
            for ch in chunk.get("choices", []):
                delta = ch.get("delta", {})
                produced = False
                if delta.get("content"):
                    content_chars += len(delta["content"])
                    produced = True
                for key in ("reasoning_content", "reasoning"):
                    if delta.get(key):
                        reasoning_chars += len(delta[key])
                        produced = True
                if delta.get("tool_calls"):
                    tool_calls.extend(delta["tool_calls"])
                    produced = True
                if produced:
                    now = time.perf_counter()
                    if t_first is None:
                        t_first = now
                    t_last = now
                if ch.get("finish_reason"):
                    finish = ch["finish_reason"]
            if chunk.get("metrics"):
                req_metrics = chunk["metrics"]
    t_end = time.perf_counter()
    if usage is None:
        raise RuntimeError(f"engine sent no usage chunk (finish={finish})")
    if t_first is None:
        raise RuntimeError(f"engine produced no output delta (finish={finish}, usage={usage})")
    completion = usage["completion_tokens"]
    decode_span = (t_last - t_first) if t_last and t_last > t_first else None
    return {
        "ttft_s": t_first - t0,
        "total_s": t_end - t0,
        "prompt_tokens": usage["prompt_tokens"],
        "completion_tokens": completion,
        "decode_tps": ((completion - 1) / decode_span) if decode_span and completion > 1 else None,
        "finish_reason": finish,
        "content_chars": content_chars,
        "reasoning_chars": reasoning_chars,
        "tool_call_deltas": len(tool_calls),
        "request_metrics": req_metrics,
    }


def measured(base: str, body: dict) -> dict:
    before = snapshot(base)
    result = stream_chat(base, body)
    after = snapshot(base)
    result.update(deltas(before, after))
    saved = result["cache"].get("tokens_saved", 0)
    uncached = result["prompt_tokens"] - saved
    result["uncached_prompt_tokens"] = uncached
    result["prefill_tps"] = uncached / result["ttft_s"] if result["ttft_s"] > 0 else None
    return result


# ---------------------------------------------------------------- fixed texts

WORDS = (
    "system memory engine model token cache request worker queue latency throughput "
    "prefill decode kernel buffer tensor layer attention state checkpoint budget window "
    "the a of to and in for on with by from at as into over under between across through "
    "reads writes measures holds returns stores computes loads keeps moves drops shares "
    "quickly slowly carefully fully partly rarely often always never exactly roughly "
    "server client network disk file record entry index table column value field schema "
    "first second third final early late fresh stale warm cold large small long short "
    "user operator developer reviewer tester maintainer owner service scheduler supervisor"
).split()


def document(n_words: int, seed: int) -> str:
    rng = random.Random(seed)
    sentences, count = [], 0
    while count < n_words:
        k = rng.randint(8, 18)
        words = [rng.choice(WORDS) for _ in range(k)]
        words[0] = words[0].capitalize()
        sentences.append(" ".join(words) + ".")
        count += k
        if rng.random() < 0.12:
            sentences.append("\n\n")
    return " ".join(sentences)


SHORT_PROMPT = (
    "You are reviewing the design of a small inference server that runs one language model on a "
    "single workstation. The server keeps a prefix cache of previously processed prompts, admits "
    "at most eight concurrent requests, and speculatively drafts a few tokens per step with a "
    "lightweight head that the full model then verifies. Operators complain that the first token "
    "of a long request can take many seconds, while short chats feel fast. Explain, in plain "
    "prose and without lists, where the time goes for a long prompt versus a short one, what the "
    "prefix cache can and cannot save when the end of every prompt changes, and which two "
    "measurements an operator should take before changing any setting. Write at least four "
    "paragraphs. Keep each paragraph concrete: name the component, the cost, and the reason, and "
    "say what you would expect to see on a machine with ample memory but a single GPU."
)

TURN_TOOL_CALLS = [
    ("tree", {"path": "/Users/operator/project/src", "depth": 2}),
    ("shell", {"command": "rg -n 'fn handle_request' /Users/operator/project/src"}),
    ("shell", {"command": "sed -n '1,160p' /Users/operator/project/src/server/handler.rs"}),
]


def tool_result(turn: int) -> str:
    rng = random.Random(1000 + turn)
    lines = []
    for i in range(140):
        indent = "    " * rng.randint(0, 3)
        words = " ".join(rng.choice(WORDS) for _ in range(rng.randint(3, 9)))
        lines.append(f"{i + 1:4d}  {indent}{words.replace(' ', '_', 1)}({rng.randint(0, 99)});")
    return "\n".join(lines)


def turn_context(turn: int, context_used: int | None, window: int) -> str:
    minute = 7 + 2 * turn
    ctx = (
        "context: no call completed yet"
        if context_used is None
        else f"context: {context_used:,} of {window:,} tokens used ({round(100 * context_used / window)}%)"
    )
    return "\n".join(
        [
            "<turn-context>",
            f"<current-time>2026-09-23 14:{minute:02d}:00</current-time>",
            "<working-directory>/Users/operator/project</working-directory>",
            f"<context>{ctx}</context>",
            "",
            "<ledger>\nThe project ledger (.goose/ledger.md) is empty — ledger_append the first finding, decision or dead end you meet.\n</ledger>",
            "</turn-context>",
        ]
    )


# ---------------------------------------------------------------- workloads


def sampling(args) -> dict:
    return {} if args.omit_temperature else {"temperature": args.temperature}


def workload_a(base, model, args, rep):
    msg = f"Session {args.nonce}-a{rep}.\n{SHORT_PROMPT}"
    return [measured(base, {"model": model, "messages": [{"role": "user", "content": msg}], "max_tokens": 300, **sampling(args)})]


def workload_b(base, model, args, rep):
    doc = document(args.long_words, seed=42)
    msg = (
        f"Session {args.nonce}-b{rep}.\nBelow is an operations log. Read it, then answer the question after it.\n\n"
        f"{doc}\n\nQuestion: in about 150 words of plain prose, which components does the log mention most, and what does it say about them?"
    )
    return [measured(base, {"model": model, "messages": [{"role": "user", "content": msg}], "max_tokens": 200, **sampling(args)})]


def workload_c(base, model, args, rep, fixture, window):
    system = f"Session {args.nonce}-c{rep}.\n" + fixture["system_prompt"]
    task = (
        "Find where the server handles an incoming request, read the handler, and tell me in two "
        "sentences where a request could wait before its first token."
    )
    history = [{"role": "system", "content": system}, {"role": "user", "content": task}]
    turns, context_used = [], None
    for turn in range(4):
        tail = turn_context(turn, context_used, window)
        messages = [dict(m) for m in history]
        if messages[-1]["role"] == "user":
            messages[-1]["content"] = messages[-1]["content"] + "\n" + tail
        else:
            messages.append({"role": "user", "content": tail})
        body = {"model": model, "messages": messages, "tools": fixture["tools"], "max_tokens": 200, **sampling(args)}
        r = measured(base, body)
        r["turn"] = turn + 1
        turns.append(r)
        context_used = r["prompt_tokens"] + r["completion_tokens"]
        if turn < len(TURN_TOOL_CALLS):
            name, arguments = TURN_TOOL_CALLS[turn]
            call_id = f"call_{turn + 1}"
            history.append(
                {
                    "role": "assistant",
                    "content": "",
                    "tool_calls": [{"id": call_id, "type": "function", "function": {"name": name, "arguments": json.dumps(arguments)}}],
                }
            )
            history.append({"role": "tool", "tool_call_id": call_id, "content": tool_result(turn)})
    return turns


# ---------------------------------------------------------------- reporting


def summarize(values):
    vals = [v for v in values if v is not None]
    if not vals:
        return None
    return {"median": statistics.median(vals), "min": min(vals), "max": max(vals), "n": len(vals)}


def fmt(s, digits=2):
    if s is None:
        return "n/a"
    return f"{s['median']:.{digits}f} ({s['min']:.{digits}f}–{s['max']:.{digits}f})"


def spec_accept(runs):
    drafted = accepted = verify = 0.0
    for r in runs:
        rm = (r.get("request_metrics") or {}).get("speculative_decoding") or {}
        drafted += sum(rm.get("drafted_by_depth") or [])
        accepted += sum(rm.get("accepted_by_depth") or [])
        verify += rm.get("verify_calls") or 0
    return drafted, accepted, verify


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--label", required=True)
    ap.add_argument("--base-url", default="http://127.0.0.1:8090")
    ap.add_argument("--reps", type=int, default=3)
    ap.add_argument("--workloads", default="a,b,c")
    ap.add_argument("--nonce", default=None)
    ap.add_argument("--temperature", type=float, default=0.0)
    ap.add_argument("--omit-temperature", action="store_true")
    ap.add_argument("--long-words", type=int, default=28600, help="document size for (b); ~32k tokens at the default (measured 24,000 words -> 26,838 tokens)")
    ap.add_argument("--rep-start", type=int, default=0, help="first repetition index (replays a specific nonce-rep)")
    args = ap.parse_args()
    base = args.base_url.rstrip("/")
    args.nonce = args.nonce or f"{args.label}-{datetime.now().strftime('%Y%m%dT%H%M%S')}"

    models = json.loads(http_get(f"{base}/v1/models"))["data"]
    model, window = models[0]["id"], models[0].get("context_window") or 0
    engine = {k: models[0].get(k) for k in ("id", "context_window", "is_hybrid", "tool_call_parser", "reasoning_parser", "speculative_decoding")}
    status0 = json.loads(http_get(f"{base}/v1/status"))
    if status0.get("num_running") or status0.get("num_waiting"):
        sys.exit(f"refusing: engine is busy ({status0.get('num_running')} running, {status0.get('num_waiting')} waiting)")
    fixture = json.load(open(os.path.join(HERE, "fixtures", "goose-agent-request.json")))

    warm = stream_chat(base, {"model": model, "messages": [{"role": "user", "content": f"Warmup {args.nonce}: reply OK."}], "max_tokens": 8, **sampling(args)})

    results = {"label": args.label, "nonce": args.nonce, "started": datetime.now().isoformat(), "engine": engine,
               "sampling": sampling(args) or "engine default (omitted)", "reps": args.reps, "warmup": warm, "workloads": {}}
    for w in args.workloads.split(","):
        runs = []
        for rep in range(args.rep_start, args.rep_start + args.reps):
            if w == "a":
                out = workload_a(base, model, args, rep)
            elif w == "b":
                out = workload_b(base, model, args, rep)
            elif w == "c":
                out = workload_c(base, model, args, rep, fixture, window)
            else:
                sys.exit(f"unknown workload {w}")
            runs.append(out)
            last = out[-1]
            print(f"[{w} rep{rep}] ttft={last['ttft_s']:.2f}s prompt={last['prompt_tokens']} out={last['completion_tokens']} "
                  f"decode={last['decode_tps'] or 0:.1f}t/s saved={last['cache'].get('tokens_saved')}", flush=True)
        results["workloads"][w] = runs
    results["finished"] = datetime.now().isoformat()
    results["status_after"] = json.loads(http_get(f"{base}/v1/status"))

    out_dir = os.path.join(HERE, "results", f"{datetime.now().strftime('%Y-%m-%d')}-{args.label}")
    os.makedirs(out_dir, exist_ok=True)
    with open(os.path.join(out_dir, "results.json"), "w") as f:
        json.dump(results, f, indent=1)

    md = [f"# {args.label}", "", f"engine `{model}` · sampling {results['sampling']} · reps {args.reps} · nonce `{args.nonce}`",
          "", "median (min–max) over reps", "",
          "| workload | TTFT s | prefill tok/s (uncached) | decode tok/s | prompt tok | completion tok | cache tokens_saved | MTP accepted/drafted (verify calls) |",
          "|---|---|---|---|---|---|---|---|"]
    for w, runs in results["workloads"].items():
        if w == "c":
            for turn in range(len(runs[0])):
                rs = [run[turn] for run in runs]
                d, a, v = spec_accept(rs)
                md.append(f"| c turn {turn + 1} | {fmt(summarize([r['ttft_s'] for r in rs]))} | {fmt(summarize([r['prefill_tps'] for r in rs]), 0)} | "
                          f"{fmt(summarize([r['decode_tps'] for r in rs]), 1)} | {fmt(summarize([r['prompt_tokens'] for r in rs]), 0)} | "
                          f"{fmt(summarize([r['completion_tokens'] for r in rs]), 0)} | {fmt(summarize([r['cache'].get('tokens_saved') for r in rs]), 0)} | "
                          f"{a:.0f}/{d:.0f} ({v:.0f}) |")
        else:
            rs = [run[0] for run in runs]
            d, a, v = spec_accept(rs)
            md.append(f"| {w} | {fmt(summarize([r['ttft_s'] for r in rs]))} | {fmt(summarize([r['prefill_tps'] for r in rs]), 0)} | "
                      f"{fmt(summarize([r['decode_tps'] for r in rs]), 1)} | {fmt(summarize([r['prompt_tokens'] for r in rs]), 0)} | "
                      f"{fmt(summarize([r['completion_tokens'] for r in rs]), 0)} | {fmt(summarize([r['cache'].get('tokens_saved') for r in rs]), 0)} | "
                      f"{a:.0f}/{d:.0f} ({v:.0f}) |")
    with open(os.path.join(out_dir, "summary.md"), "w") as f:
        f.write("\n".join(md) + "\n")
    print("\n".join(md))
    print(f"\nwritten: {out_dir}")


if __name__ == "__main__":
    main()

File: crates/goose-sidecar/src/placement/bench.rs
//! "Measure speed": a fixed, token-counted workload sent to a running engine — a ~2k-token prompt
//! then up to 256 greedy tokens (the shape of our recorded runs, `mlx_lm.benchmark -p 2048 -g 256`),
//! and optionally a ~32k-token document for the long-documents goal. Nothing here is timed out or
//! capped by a clock: the workload ends when the engine finishes the tokens it was asked for.
//!
//! The prompt is the eval bench's document generator (evals/mlx-engine-bench/bench.py `document`,
//! the same vocabulary) opened by a NONCE, so the prefix cache cannot serve a previous run's prompt
//! and the time to the first token is a cold prefill.

use std::time::Instant;

use anyhow::{bail, Context, Result};
use serde::{Deserialize, Serialize};

/// bench.py's WORDS: the vocabulary its fixed documents are made of.
const WORDS: &[&str] = &[
    "system",
    "memory",
    "engine",
    "model",
    "token",
    "cache",
    "request",
    "worker",
    "queue",
    "latency",
    "throughput",
    "prefill",
    "decode",
    "kernel",
    "buffer",
    "tensor",
    "layer",
    "attention",
    "state",
    "checkpoint",
    "budget",
    "window",
    "the",
    "a",
    "of",
    "to",
    "and",
    "in",
    "for",
    "on",
    "with",
    "by",
    "from",
    "at",
    "as",
    "into",
    "over",
    "under",
    "between",
    "across",
    "through",
    "reads",
    "writes",
    "measures",
    "holds",
    "returns",
    "stores",
    "computes",
    "loads",
    "keeps",
    "moves",
    "drops",
    "shares",
    "quickly",
    "slowly",
    "carefully",
    "fully",
    "partly",
    "rarely",
    "often",
    "always",
    "never",
    "exactly",
    "roughly",
    "server",
    "client",
    "network",
    "disk",
    "file",
    "record",
    "entry",
    "index",
    "table",
    "column",
    "value",
    "field",
    "schema",
    "first",
    "second",
    "third",
    "final",
    "early",
    "late",
    "fresh",
    "stale",
    "warm",
    "cold",
    "large",
    "small",
    "long",
    "short",
    "user",
    "operator",
    "developer",
    "reviewer",
    "tester",
    "maintainer",
    "owner",
    "service",
    "scheduler",
    "supervisor",
];

// measured: bench.py's document of 24,000 of these words tokenized to 26,838 Qwen tokens
// (1.118 tokens per word); prompts are sized in words by this ratio and the engine's own
// `usage.prompt_tokens` is what gets recorded.
const TOKENS_PER_WORD: f64 = 26_838.0 / 24_000.0;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Workload {
    /// ~2k prompt tokens, 256 generated: the chat goal's decode and a short prefill.
    Chat,
    /// ~30k prompt tokens, a short answer: the long-documents goal's prefill.
    LongDocument,
}

impl Workload {
    /// Prompt tokens aimed for (the document is sized to land just under the bucket).
    pub fn prompt_tokens(self) -> u64 {
        match self {
            // ratio: the recorded runs' 2,048-token prompt, less room for the instruction.
            Workload::Chat => 1_900,
            // ratio: bench.py's (b) long document class (~32k), less room for the instruction.
            Workload::LongDocument => 30_000,
        }
    }

    pub fn max_tokens(self) -> u32 {
        match self {
            // ratio: the recorded runs' `-g 256`.
            Workload::Chat => 256,
            // ratio: bench.py (b)'s answer is 200 tokens; prefill is the figure this pass is for.
            Workload::LongDocument => 64,
        }
    }

    /// The answer a typical turn of this shape writes — what a whole turn is timed with when goose
    /// has no recorded turns of the model at this size. (The measurement pass stops at
    /// `max_tokens`: it is after the prompt's rate, not the answer.)
    pub fn answer_tokens(self) -> u64 {
        match self {
            // ratio: the recorded runs' `-g 256`.
            Workload::Chat => 256,
            // ratio: bench.py (b)'s long-document answer is 200 tokens.
            Workload::LongDocument => 200,
        }
    }

    /// The context bucket the workload's prompt falls in.
    pub fn bucket(self) -> u64 {
        super::store::context_bucket(self.prompt_tokens())
    }

    /// Context a placement needs to run this workload at all: its prompt plus its answer.
    pub fn context_needed(self) -> u64 {
        self.bucket() + self.max_tokens() as u64
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Workload::Chat => "chat",
            Workload::LongDocument => "longDocument",
        }
    }
}

/// xorshift64*: a fixed document from a seed without a dependency.
struct Rng(u64);

impl Rng {
    fn next(&mut self) -> u64 {
        self.0 ^= self.0 >> 12;
        self.0 ^= self.0 << 25;
        self.0 ^= self.0 >> 27;
        self.0.wrapping_mul(0x2545_F491_4F6C_DD1D)
    }

    fn below(&mut self, n: u64) -> u64 {
        self.next() % n
    }
}

/// bench.py's `document`: sentences of 8–18 words, a paragraph break now and then.
fn document(words: usize, seed: u64) -> String {
    let mut rng = Rng(seed.max(1));
    let mut out = String::new();
    let mut count = 0;
    while count < words {
        let k = 8 + rng.below(11) as usize;
        let mut sentence: Vec<String> = (0..k)
            .map(|_| WORDS[rng.below(WORDS.len() as u64) as usize].to_string())
            .collect();
        if let Some(first) = sentence.first_mut() {
            let mut chars = first.chars();
            if let Some(c) = chars.next() {
                *first = c.to_uppercase().collect::<String>() + chars.as_str();
            }
        }
        out.push_str(&sentence.join(" "));
        out.push_str(". ");
        count += k;
        if rng.below(100) < 12 {
            out.push_str("\n\n");
        }
    }
    out
}

/// The workload's prompt, opened by `nonce`.
pub fn prompt(workload: Workload, nonce: &str) -> String {
    let words = (workload.prompt_tokens() as f64 / TOKENS_PER_WORD) as usize;
    let instruction = match workload {
        Workload::Chat => {
            "Continue the operations log below in the same style, one sentence after another, \
             without lists or headings. Keep writing until you are stopped."
        }
        Workload::LongDocument => {
            "Read the operations log below, then name in one short paragraph the three components \
             it mentions most often."
        }
    };
    format!(
        "Measurement {nonce}.\n{instruction}\n\n{}",
        document(words, 0x9E37_79B9 ^ workload.prompt_tokens())
    )
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BenchSample {
    pub workload: Workload,
    pub prompt_tokens: u64,
    /// Prompt tokens the engine said its cache served (`usage.prompt_tokens_details.cached_tokens`).
    pub cached_tokens: u64,
    pub completion_tokens: u64,
    pub ttft_ms: f64,
    /// (prompt − cached) ÷ time to first token — conservative: the first token's own step is in it.
    pub prefill_tps: f64,
    /// (completion − 1) ÷ the span between the first and the last output delta; `None` when the
    /// engine produced one token or one delta (no span to divide).
    pub decode_tps: Option<f64>,
    pub finish_reason: Option<String>,
}

fn is_output_delta(choice: &serde_json::Value) -> bool {
    let delta = &choice["delta"];
    ["content", "reasoning_content", "reasoning"]
        .iter()
        .any(|k| delta[k].as_str().is_some_and(|s| !s.is_empty()))
        || delta["tool_calls"]
            .as_array()
            .is_some_and(|a| !a.is_empty())
}

/// Stream one workload request through `base_url`'s `/v1/chat/completions` and time it.
pub async fn run_workload(
    base_url: &str,
    served_model: &str,
    workload: Workload,
    nonce: &str,
) -> Result<BenchSample> {
    let body = serde_json::json!({
        "model": served_model,
        "messages": [{"role": "user", "content": prompt(workload, nonce)}],
        "max_tokens": workload.max_tokens(),
        "temperature": 0,
        "stream": true,
        "stream_options": {"include_usage": true},
    });
    let url = format!("{}/v1/chat/completions", base_url.trim_end_matches('/'));
    let started = Instant::now();
    let mut response = reqwest::Client::new()
        .post(&url)
        .header(reqwest::header::CONTENT_TYPE, "application/json")
        .body(serde_json::to_vec(&body)?)
        .send()
        .await
        .with_context(|| format!("POST {url}"))?;
    let status = response.status();
    if !status.is_success() {
        let text = response.text().await.unwrap_or_default();
        bail!("{url} answered {status}: {}", text.trim());
    }
    let mut buffer = String::new();
    let (mut first, mut last): (Option<Instant>, Option<Instant>) = (None, None);
    let mut usage: Option<serde_json::Value> = None;
    let mut finish_reason = None;
    let mut done = false;
    while let Some(chunk) = response.chunk().await.context("reading the stream")? {
        buffer.push_str(&String::from_utf8_lossy(&chunk));
        while let Some(end) = buffer.find('\n') {
            let line: String = buffer.drain(..=end).collect();
            let Some(payload) = line.trim().strip_prefix("data:").map(str::trim) else {
                continue;
            };
            if payload == "[DONE]" {
                done = true;
                continue;
            }
            let event: serde_json::Value = serde_json::from_str(payload)
                .with_context(|| format!("an SSE event that is not JSON: {payload}"))?;
            if let Some(error) = event.get("error") {
                bail!("the engine streamed an error: {error}");
            }
            if event["usage"].is_object() {
                usage = Some(event["usage"].clone());
            }
            for choice in event["choices"].as_array().into_iter().flatten() {
                if is_output_delta(choice) {
                    let now = Instant::now();
                    first.get_or_insert(now);
                    last = Some(now);
                }
                if let Some(reason) = choice["finish_reason"].as_str() {
                    finish_reason = Some(reason.to_string());
                }
            }
        }
    }
    if !done {
        bail!("the stream ended without [DONE] (finish reason {finish_reason:?})");
    }
    let usage = usage.context("the engine sent no usage chunk")?;
    let first = first.context("the engine produced no output token")?;
    let tokens = |key: &str| {
        usage[key]
            .as_u64()
            .with_context(|| format!("usage has no `{key}`: {usage}"))
    };
    let prompt_tokens = tokens("prompt_tokens")?;
    let completion_tokens = tokens("completion_tokens")?;
    // An engine that does not report cache hits served none of THIS prompt: its nonce sits at
    // token 0, so no cached prefix can match past the chat template's header.
    let cached_tokens = usage["prompt_tokens_details"]["cached_tokens"]
        .as_u64()
        .unwrap_or(0);
    let ttft_ms = first.duration_since(started).as_secs_f64() * 1000.0;
    let span_s = last.map(|l| l.duration_since(first).as_secs_f64());
    Ok(BenchSample {
        workload,
        prompt_tokens,
        cached_tokens,
        completion_tokens,
        ttft_ms,
        prefill_tps: prompt_tokens.saturating_sub(cached_tokens) as f64 / (ttft_ms / 1000.0),
        decode_tps: span_s
            .filter(|s| *s > 0.0 && completion_tokens > 1)
            .map(|s| (completion_tokens - 1) as f64 / s),
        finish_reason,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::io::{AsyncReadExt, AsyncWriteExt};

    #[test]
    fn the_prompt_is_fixed_but_its_nonce_is_not() {
        let a = prompt(Workload::Chat, "n1");
        assert_eq!(
            a,
            prompt(Workload::Chat, "n1"),
            "the document is deterministic"
        );
        assert_ne!(a, prompt(Workload::Chat, "n2"));
        assert!(a.starts_with("Measurement n1."));
        let words = a.split_whitespace().count() as f64;
        let tokens = words * TOKENS_PER_WORD;
        assert!(tokens < 2048.0 && tokens > 1700.0, "{tokens}");
        assert_eq!(Workload::Chat.bucket(), 2048);
        assert_eq!(Workload::LongDocument.bucket(), 32_768);
        assert_eq!(Workload::Chat.context_needed(), 2048 + 256);
    }

    /// A one-shot SSE server: what the engine streams, byte for byte.
    async fn serve_once(body: &'static str) -> String {
        let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();
        tokio::spawn(async move {
            let (mut socket, _) = listener.accept().await.unwrap();
            let mut request = vec![0u8; 65_536];
            let _ = socket.read(&mut request).await;
            let head = format!(
                "HTTP/1.1 200 OK\r\ncontent-type: text/event-stream\r\ncontent-length: {}\r\n\r\n",
                body.len()
            );
            socket.write_all(head.as_bytes()).await.unwrap();
            socket.write_all(body.as_bytes()).await.unwrap();
        });
        format!("http://{addr}")
    }

    #[tokio::test]
    async fn a_stream_is_timed_from_its_own_usage_and_deltas() {
        let base = serve_once(concat!(
            "data: {\"choices\":[{\"delta\":{\"role\":\"assistant\"}}]}\n\n",
            "data: {\"choices\":[{\"delta\":{\"content\":\"a\"}}]}\n\n",
            "data: {\"choices\":[{\"delta\":{\"content\":\"b\"},\"finish_reason\":\"length\"}]}\n\n",
            "data: {\"choices\":[],\"usage\":{\"prompt_tokens\":1900,\"completion_tokens\":2,\"prompt_tokens_details\":{\"cached_tokens\":4}}}\n\n",
            "data: [DONE]\n\n"
        ))
        .await;
        let sample = run_workload(&base, "m", Workload::Chat, "x").await.unwrap();
        assert_eq!(sample.prompt_tokens, 1900);
        assert_eq!(sample.cached_tokens, 4);
        assert_eq!(sample.completion_tokens, 2);
        assert_eq!(sample.finish_reason.as_deref(), Some("length"));
        assert!(sample.prefill_tps > 0.0);
    }

    #[tokio::test]
    async fn a_stream_without_usage_or_done_is_a_named_failure() {
        let base =
            serve_once("data: {\"choices\":[{\"delta\":{\"content\":\"a\"}}]}\n\ndata: [DONE]\n\n")
                .await;
        let err = format!(
            "{:#}",
            run_workload(&base, "m", Workload::Chat, "x")
                .await
                .unwrap_err()
        );
        assert!(err.contains("no usage chunk"), "{err}");
        let base = serve_once("data: {\"choices\":[{\"delta\":{\"content\":\"a\"}}]}\n\n").await;
        let err = format!(
            "{:#}",
            run_workload(&base, "m", Workload::Chat, "x")
                .await
                .unwrap_err()
        );
        assert!(err.contains("without [DONE]"), "{err}");
    }
}
